ops/main.py

```diff
@@ -42,6 +43,87 @@
 logger = logging.getLogger()
 
 
+@dataclasses.dataclass(frozen=True)
+class _JujuContext:
+    # source: https://juju.is/docs/juju/charm-environment-variables
+    # the HookVars function: https://github.com/juju/juju/blob/3.6/worker/uniter/runner/context/context.go#L1398
+    # only allow accesing these vars because these are the currently used vars in main.py
+
+    juju_action: '_JujuAction'
+    juju_charm_dir: Optional[str]
+    juju_debug: Optional[str]
+    juju_departing_unit: str
+    juju_dispatch_path: Optional[str]
+    juju_notice: '_JujuNotice'
+    juju_pebble_check_name: str
```

Comment:
I imagine this too should be scoped under... a check? a container? a notice? Ugh this is getting difficult...

Reply:
It indeed gets messy. It shouldn't really be under notice because notices are meant to be an implementation detail of check events. It's really check and notice both under container, but I don't think we want to be three levels deep (I'm not even convinced of two levels).

Reply:
Yeah, I'm with Tony on this one, and this is part of the reason that hierarchies -- while seeming enticing at first -- are often not ideal. Sometimes it's unclear which hierarchy something belongs in, and sometimes a thing belongs in multiple hierarchies.

In addition, if we're going for nested, I don't like how we have to have all these different classes, like `_JujuRelation` (and it's actually not a relation, it's an "extended relation identifier" or something).

So I'm much more in favour of flat, 1-level for this, as #1313 does.

Reply:
Additionally, we convert to "proper" types further down, like converting notice id/type/key to `LazyNotice` in the notice events. I think that's another reason to keep the types here flat and simple.